File: elements-of-programming-interviews/cpp/do_terminated_lists_overlap.cc
#include <memory>

#include "list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using JoshuaListTools::PrintList;
using std::shared_ptr;

/*

Approach 1 (not implemented). O(n) time, O(c) space, but modifies list.
- Measure length of both
- Reverse both twice
- Remeasure lengths:
  - If lengths change, they overlap (one of them will be shorter because
    reversal will break joint)
- Walk both from heads; when shorter's next is null, set it to other's next.
  Then return that next.

Insight: If two lists are the same length and overlap, then they both meet at
their ith node (the shared portion will be the same length for both, so they can
only have the same overall length if their unshared portions are equally as
long).

Approach 2 (implemented). O(n) time, O(c) space, no side effects.
We only need to tell if two lists overlap, and if so we can return any node
in them. They're both sure to terminate, so why not just iterate to the end
of both and compare the address of the final nodes?

Unfortunately the book doesn't say this, but the test cases expect you to
return the first node they have in common, not any of them.


Approach 3 (implemented). O(n) time, O(c) space, no side effects.
- Measure length of both
- If lengths are unequal, advance the head of the longer one until they are.
- Advance heads one at a time until a joint node is found (or end is reached)
*/

shared_ptr<ListNode<int>> OverlappingNoCycleListsAnyNode(
    shared_ptr<ListNode<int>> l0, shared_ptr<ListNode<int>> l1) {
  if (!l0 || !l1) {
    return nullptr;
  }
  PrintList(l0);
  PrintList(l1);

  while (l0->next != nullptr) {
    l0 = l0->next;
  };
  while (l1->next != nullptr) {
    l1 = l1->next;
  };

  return (l0.get() == l1.get()) ? l0 : nullptr;
}

shared_ptr<ListNode<int>> OverlappingNoCycleLists(
    shared_ptr<ListNode<int>> l0, shared_ptr<ListNode<int>> l1) {
  if (!l0 || !l1) {
    return nullptr;
  }
  int l0Length = JoshuaListTools::MeasureListLength(l0),
      l1Length = JoshuaListTools::MeasureListLength(l1);
  while (l0Length > l1Length) {
    l0 = l0->next;
    l0Length--;
  }
  while (l1Length > l0Length) {
    l1 = l1->next;
    l1Length--;
  }

  while (l0 && l1) {
    if (l0 == l1) {
      return l0;
    }
    l0 = l0->next;
    l1 = l1->next;
  }
  return nullptr;
}

void OverlappingNoCycleListsWrapper(TimedExecutor& executor,
                                    shared_ptr<ListNode<int>> l0,
                                    shared_ptr<ListNode<int>> l1,
                                    shared_ptr<ListNode<int>> common) {
  if (common) {
    if (l0) {
      auto i = l0;
      while (i->next) {
        i = i->next;
      }
      i->next = common;
    } else {
      l0 = common;
    }

    if (l1) {
      auto i = l1;
      while (i->next) {
        i = i->next;
      }
      i->next = common;
    } else {
      l1 = common;
    }
  }

  auto result = executor.Run([&] { return OverlappingNoCycleLists(l0, l1); });

  if (result != common) {
    printf("Invalid result; expected: %p (%d), got %p (%d)\n", common.get(),
           common->data, result.get(), result->data);
    throw TestFailure("Test failed.\n");
  }
}

int main(int argc, char* argv[]) {
  /*
  shared_ptr<ListNode<int>> head = ConstructList(5);
  PrintList(head);
  auto newHead = ReverseList(head);
  PrintList(newHead);
  */

  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "l0", "l1", "common"};
  return GenericTestMain(
      args, "do_terminated_lists_overlap.cc", "do_terminated_lists_overlap.tsv",
      &OverlappingNoCycleListsWrapper, DefaultComparator{}, param_names);
}

File: elements-of-programming-interviews/2023/python/do_terminated_lists_overlap.py
import functools

from list_node import ListNode
from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook

# if lists are same length and overlap, they overlap at the same place
# so shorten one if necessary, then compare node-by-node.


def get_list_len(head: ListNode):
    size = 0
    while head:
        size += 1
        head = head.next
    return size


def overlapping_no_cycle_lists(l0: ListNode, l1: ListNode) -> ListNode:
    l0_len = get_list_len(l0)
    l1_len = get_list_len(l1)
    while l0_len > l1_len:
        l0 = l0.next
        l0_len -= 1
    while l0_len < l1_len:
        l1 = l1.next
        l1_len -= 1

    while l0 is not l1:
        l0 = l0.next
        l1 = l1.next

    return l0 if l0 is l1 and l0 is not None else None


@enable_executor_hook
def overlapping_no_cycle_lists_wrapper(executor, l0, l1, common):
    if common:
        if l0:
            i = l0
            while i.next:
                i = i.next
            i.next = common
        else:
            l0 = common

        if l1:
            i = l1
            while i.next:
                i = i.next
            i.next = common
        else:
            l1 = common

    result = executor.run(functools.partial(overlapping_no_cycle_lists, l0,
                                            l1))

    if result != common:
        raise TestFailure('Invalid result')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('do_terminated_lists_overlap.py',
                                       'do_terminated_lists_overlap.tsv',
                                       overlapping_no_cycle_lists_wrapper))
